src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.GateThread.cs

```diff
@@ -13,70 +13,74 @@ private static class GateThread
+                                    while (counts.numExistingThreads < ThreadPoolInstance._maxThreads && counts.numExistingThreads >= counts.numThreadsGoal)
                                     {
-                                        Debug.WriteLine("The CLR ThreadPool detected work starvation!");
-                                        Debugger.Break();
-                                    }
+                                        if (debuggerBreakOnWorkStarvation)
+                                        {
+                                            Debugger.Break();
+                                        }
 
-                                    ThreadCounts newCounts = counts;
-                                    newCounts.numThreadsGoal = (short)(newCounts.numExistingThreads + 1);
-                                    ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref ThreadPoolInstance._separated.counts, newCounts, counts);
-                                    if (oldCounts == counts)
-                                    {
-                                        HillClimbing.ThreadPoolHillClimber.ForceChange(newCounts.numThreadsGoal, HillClimbing.StateOrTransition.Starvation);
-                                        WorkerThread.MaybeAddWorkingWorker();
-                                        break;
+                                        ThreadCounts newCounts = counts;
+                                        newCounts.numThreadsGoal = (short)(newCounts.numExistingThreads + 1);
+                                        ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref ThreadPoolInstance._separated.counts, newCounts, counts);
+                                        if (oldCounts == counts)
+                                        {
+                                            HillClimbing.ThreadPoolHillClimber.ForceChange(newCounts.numThreadsGoal, HillClimbing.StateOrTransition.Starvation);
+                                            WorkerThread.MaybeAddWorkingWorker();
+                                            break;
+                                        }
+                                        counts = oldCounts;
                                     }
-                                    counts = oldCounts;
                                 }
-                            }
-                            finally
-                            {
-                                ThreadPoolInstance._hillClimbingThreadAdjustmentLock.Release();
+                                finally
+                                {
+                                    ThreadPoolInstance._hillClimbingThreadAdjustmentLock.Release();
+                                }
                             }
                         }
-                    }
+                    } while (Interlocked.Decrement(ref numRunsRemaining) > GetRunningMaskForNumRuns(0));
```

Comment:
The gate thread shouldn't go inactive when there are pending thread requests because that is typically when the starvation heuristic needs to trigger